dropshot/src/extractor/body.rs

```diff
@@ -184,38 +189,246 @@ impl UntypedBody {
+        self.into_stream()
+            .try_fold(BytesMut::new(), |mut out, chunk| {
+                out.put(chunk);
+                futures::future::ok(out)
+            })
+            .await
+    }
+
+    /// Converts `self` into a stream.
+    ///
+    /// The `Stream` produces values of type `Result<Bytes, HttpError>`.
+    ///
+    /// # Errors
+    ///
+    /// The stream produces an [`HttpError`] if any of the following cases occur:
+    ///
+    /// * A network error occurred.
+    /// * `request_body_max_bytes` was exceeded for this request.
+    ///
+    /// # Examples
+    ///
+    /// Buffer a `StreamingBody` in-memory, into a
+    /// [`BufList`](https://docs.rs/buf-list/latest/buf_list/struct.BufList.html)
+    /// (a segmented list of [`Bytes`] chunks). This is similar to
+    /// [`Self::into_bytes_mut`], except it avoids copying memory into a single
+    /// large allocation.
+    ///
+    /// ```
+    /// use buf_list::BufList;
+    /// use dropshot::{HttpError, StreamingBody};
+    /// use futures::prelude::*;
+    /// # use std::iter::FromIterator;
+    ///
+    /// async fn into_buf_list(body: StreamingBody) -> Result<BufList, HttpError> {
+    ///     body.into_stream().try_collect().await
+    /// }
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// #    let body = StreamingBody::__from_bytes(bytes::Bytes::from("foobar"));
+    /// #    assert_eq!(
+    /// #        into_buf_list(body).await.unwrap().into_iter().next(),
+    /// #        Some(bytes::Bytes::from("foobar")),
+    /// #    );
+    /// # }
+    /// ```
+    ///
+    /// ---
+    ///
+    /// Write a `StreamingBody` to an [`AsyncWrite`](tokio::io::AsyncWrite),
+    /// for example a [`tokio::fs::File`], without buffering it into memory:
+    ///
+    /// ```
+    /// use dropshot::{HttpError, StreamingBody};
+    /// use futures::prelude::*;
+    /// use tokio::io::{AsyncWrite, AsyncWriteExt};
+    ///
+    /// async fn write_all<W: AsyncWrite + Unpin>(
+    ///     body: StreamingBody,
+    ///     writer: &mut W,
+    /// ) -> Result<(), HttpError> {
+    ///     let stream = body.into_stream();
+    ///     tokio::pin!(stream);
+    ///
+    ///     while let Some(res) = stream.next().await {
+    ///         let mut data = res?;
+    ///         writer.write_all_buf(&mut data).await.map_err(|error| {
+    ///             HttpError::for_unavail(None, format!("write failed: {error}"))
+    ///         })?;
+    ///     }
+    ///
+    ///     Ok(())
+    /// }
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// #    let body = StreamingBody::__from_bytes(bytes::Bytes::from("foobar"));
+    /// #    let mut writer = vec![];
+    /// #    write_all(body, &mut writer).await.unwrap();
+    /// #    assert_eq!(writer, &b"foobar"[..]);
+    /// # }
+    /// ```
+    ///
+    /// ---
+    ///
+    /// An alternative way to write data to an `AsyncWrite`, using
```

Comment:
What's the advantage of this?  (just thinking: did we pull in an extra dependency just to add an example that looks equivalent to the previous example?  That's not bad but it'd be useful to know why this is better.)

Reply:
That is a very good question. I figured we'd want to include this example for completeness, but I had a chat with Eliza Weisman who says that the tokio codec stuff is mostly historical:

1. Codecs were originally a way to guarantee cancel safety, but `write_all_buf` is just a much clearer way to do that -- if it fails or gets canceled, it still advances the buffer with the amount written by.
2. Back in the day the tokio devs assumed people would use `Sink` a lot more. That turned out not to be the case so much.

These days there's no reason for anyone to use codecs or `Sink`, really, so I'm going to just remove this example.

Reply:
Removed this example.